File: native/Sources/EnglishCoachCore/Engines.swift
import Foundation

public enum CourseRouting {
    public static func nextLesson(in lessons: [Lesson], completed: Set<String>) -> Lesson? {
        lessons.first { !completed.contains($0.id) }
    }

    public static func isUnlocked(index: Int, lessons: [Lesson], completed: Set<String>) -> Bool {
        guard lessons.indices.contains(index) else { return false }
        return index == 0 || completed.contains(lessons[index - 1].id)
    }
}

public enum LevelOrder {
    public static let all: [CEFRLevel] = [.a1, .a2, .b1, .b2, .c1]
    public static func next(after level: CEFRLevel) -> CEFRLevel? {
        guard let i = all.firstIndex(of: level), i + 1 < all.count else { return nil }
        return all[i + 1]
    }
}

/// Turns placement-test answers into a recommended starting level.
public enum PlacementScorer {
    /// Places the learner at the first level they cannot comfortably pass (accuracy < threshold).
    /// If they pass every level, recommends the top level; if they fail the easiest, recommends it.
    public static func recommend(bank: [PlacementQuestion], correctIDs: Set<String>, passThreshold: Double = 0.6) -> CEFRLevel {
        for level in LevelOrder.all {
            let questions = bank.filter { $0.level == level }
            guard !questions.isEmpty else { continue }
            let correct = questions.filter { correctIDs.contains($0.id) }.count
            if Double(correct) / Double(questions.count) < passThreshold { return level }
        }
        return LevelOrder.all.last ?? .a1
    }

    /// True once the answers so far already fix the recommendation, so the remaining
    /// (harder) questions cannot change it. Lets the test stop early instead of
    /// walking the learner through the whole bank.
    public static func isDecided(bank: [PlacementQuestion], answeredIDs: Set<String>, correctIDs: Set<String>, passThreshold: Double = 0.6) -> Bool {
        for level in LevelOrder.all {
            let questions = bank.filter { $0.level == level }
            guard !questions.isEmpty else { continue }
            guard questions.allSatisfy({ answeredIDs.contains($0.id) }) else { return false }
            let correct = questions.filter { correctIDs.contains($0.id) }.count
            if Double(correct) / Double(questions.count) < passThreshold { return true }
        }
        return true
    }
}

/// Level-level progress, mastery detection and adaptive difficulty helpers.
public enum ProgressionEngine {
    public static func lessons(for level: CEFRLevel, in courses: [CoursePack]) -> [Lesson] {
        courses.first { $0.level == level }?.chapters.flatMap(\.lessons) ?? []
    }

    public static func levelProgress(level: CEFRLevel, courses: [CoursePack], completed: Set<String>) -> Double {
        let all = lessons(for: level, in: courses)
        guard !all.isEmpty else { return 0 }
        return Double(all.filter { completed.contains($0.id) }.count) / Double(all.count)
    }

    public static func isLevelComplete(level: CEFRLevel, courses: [CoursePack], completed: Set<String>) -> Bool {
        let all = lessons(for: level, in: courses)
        return !all.isEmpty && all.allSatisfy { completed.contains($0.id) }
    }

    public static func exerciseIDs(for level: CEFRLevel, in courses: [CoursePack]) -> Set<String> {
        Set(lessons(for: level, in: courses).flatMap(\.exercises).map(\.id))
    }

    /// Accuracy over the most recent attempts on this level's exercises.
    public static func accuracy(level: CEFRLevel, courses: [CoursePack], attempts: [AttemptRecord], lastN: Int = 40) -> Double {
        let ids = exerciseIDs(for: level, in: courses)
        let relevant = attempts.filter { ids.contains($0.exerciseID) }.suffix(lastN)
        guard !relevant.isEmpty else { return 0 }
        return Double(relevant.filter(\.correct).count) / Double(relevant.count)
    }

    /// True when the level is fully completed with high accuracy and a harder level exists,
    /// unless the user already dismissed the suggestion for this level.
    public static func shouldSuggestAdvance(level: CEFRLevel, courses: [CoursePack], completed: Set<String>, attempts: [AttemptRecord], dismissed: Set<String>, accuracyThreshold: Double = 0.8) -> Bool {
        guard LevelOrder.next(after: level) != nil else { return false }
        guard !dismissed.contains(level.rawValue) else { return false }
        guard isLevelComplete(level: level, courses: courses, completed: completed) else { return false }
        return accuracy(level: level, courses: courses, attempts: attempts) >= accuracyThreshold
    }
}

/// Daily practice time, so the goal picked during onboarding actually means something.
public enum PracticeLog {
    /// A single very long stretch is almost always a window left open, not study time.
    public static let sessionCapSeconds = 30 * 60

    public static func dayKey(_ date: Date, calendar: Calendar = .current) -> String {
        let parts = calendar.dateComponents([.year, .month, .day], from: date)
        return String(format: "%04d-%02d-%02d", parts.year ?? 0, parts.month ?? 0, parts.day ?? 0)
    }

    public static func adding(seconds: Int, to log: [String: Int]?, on date: Date, calendar: Calendar = .current) -> [String: Int] {
        var updated = log ?? [:]
        let clamped = max(0, min(seconds, sessionCapSeconds))
        guard clamped > 0 else { return updated }
        updated[dayKey(date, calendar: calendar), default: 0] += clamped
        return updated
    }

    public static func minutes(in log: [String: Int]?, on date: Date, calendar: Calendar = .current) -> Int {
        (log?[dayKey(date, calendar: calendar)] ?? 0) / 60
    }
}

public enum ReviewEngine {
    public static func newItem(exerciseID: String, now: Date) -> ReviewItem {
        ReviewItem(id: exerciseID, exerciseID: exerciseID, due: now, intervalDays: 0, ease: 2.3, repetitions: 0)
    }

    public static func recordSuccess(_ source: ReviewItem, now: Date) -> ReviewItem {
        var item = source
        item.repetitions += 1
        if item.repetitions == 1 { item.intervalDays = 1 }
        else if item.repetitions == 2 { item.intervalDays = 3 }
        else if item.repetitions == 3 { item.intervalDays = 7 }
        else { item.intervalDays = min(180, Int((Double(max(1, item.intervalDays)) * item.ease).rounded())) }
        item.ease = min(3, item.ease + 0.05)
        item.due = Calendar.current.date(byAdding: .day, value: item.intervalDays, to: now) ?? now
        return item
    }

    public static func recordFailure(_ source: ReviewItem, now: Date) -> ReviewItem {
        var item = source
        item.repetitions = 0; item.intervalDays = 1; item.ease = max(1.6, item.ease - 0.2)
        item.due = Calendar.current.date(byAdding: .day, value: 1, to: now) ?? now
        return item
    }
}

File: web/src/core/engines.ts
import { LEVELS } from './types'
import type { AttemptRecord, CEFRLevel, CoursePack, Lesson, PlacementQuestion, ReviewItem } from './types'

/** Mirrors EnglishCoachCore/Engines.swift. Pure functions, no storage, no UI. */

function addDays(date: Date, days: number): Date {
  const result = new Date(date.getTime())
  result.setDate(result.getDate() + days)
  return result
}

// MARK: - Routing

export const CourseRouting = {
  nextLesson(lessons: Lesson[], completed: Set<string>): Lesson | null {
    return lessons.find((lesson) => !completed.has(lesson.id)) ?? null
  },

  isUnlocked(index: number, lessons: Lesson[], completed: Set<string>): boolean {
    if (index < 0 || index >= lessons.length) return false
    return index === 0 || completed.has(lessons[index - 1].id)
  },
}

export const LevelOrder = {
  all: LEVELS,
  next(after: CEFRLevel): CEFRLevel | null {
    const index = LEVELS.indexOf(after)
    return index >= 0 && index + 1 < LEVELS.length ? LEVELS[index + 1] : null
  },
}

// MARK: - Placement

export const PlacementScorer = {
  /** Places the learner at the first level they cannot comfortably pass. */
  recommend(bank: PlacementQuestion[], correctIDs: Set<string>, passThreshold = 0.6): CEFRLevel {
    for (const level of LEVELS) {
      const questions = bank.filter((q) => q.level === level)
      if (questions.length === 0) continue
      const correct = questions.filter((q) => correctIDs.has(q.id)).length
      if (correct / questions.length < passThreshold) return level
    }
    return LEVELS[LEVELS.length - 1]
  },

  /** True once the answers so far fix the recommendation, so the test can stop early. */
  isDecided(bank: PlacementQuestion[], answeredIDs: Set<string>, correctIDs: Set<string>, passThreshold = 0.6): boolean {
    for (const level of LEVELS) {
      const questions = bank.filter((q) => q.level === level)
      if (questions.length === 0) continue
      if (!questions.every((q) => answeredIDs.has(q.id))) return false
      const correct = questions.filter((q) => correctIDs.has(q.id)).length
      if (correct / questions.length < passThreshold) return true
    }
    return true
  },
}

// MARK: - Level progress

export const ProgressionEngine = {
  lessons(level: CEFRLevel, courses: CoursePack[]): Lesson[] {
    const course = courses.find((c) => c.level === level)
    return course ? course.chapters.flatMap((chapter) => chapter.lessons) : []
  },

  levelProgress(level: CEFRLevel, courses: CoursePack[], completed: Set<string>): number {
    const all = this.lessons(level, courses)
    if (all.length === 0) return 0
    return all.filter((lesson) => completed.has(lesson.id)).length / all.length
  },

  isLevelComplete(level: CEFRLevel, courses: CoursePack[], completed: Set<string>): boolean {
    const all = this.lessons(level, courses)
    return all.length > 0 && all.every((lesson) => completed.has(lesson.id))
  },

  exerciseIDs(level: CEFRLevel, courses: CoursePack[]): Set<string> {
    return new Set(this.lessons(level, courses).flatMap((lesson) => lesson.exercises).map((e) => e.id))
  },

  /** Accuracy over the most recent attempts on this level's exercises. */
  accuracy(level: CEFRLevel, courses: CoursePack[], attempts: AttemptRecord[], lastN = 40): number {
    const ids = this.exerciseIDs(level, courses)
    const relevant = attempts.filter((a) => ids.has(a.exerciseID)).slice(-lastN)
    if (relevant.length === 0) return 0
    return relevant.filter((a) => a.correct).length / relevant.length
  },

  shouldSuggestAdvance(
    level: CEFRLevel,
    courses: CoursePack[],
    completed: Set<string>,
    attempts: AttemptRecord[],
    dismissed: Set<string>,
    accuracyThreshold = 0.8,
  ): boolean {
    if (LevelOrder.next(level) === null) return false
    if (dismissed.has(level)) return false
    if (!this.isLevelComplete(level, courses, completed)) return false
    return this.accuracy(level, courses, attempts) >= accuracyThreshold
  },
}

// MARK: - Daily practice time

export const PracticeLog = {
  /** A single very long stretch is almost always a forgotten tab, not study time. */
  sessionCapSeconds: 30 * 60,

  dayKey(date: Date): string {
    const year = date.getFullYear()
    const month = `${date.getMonth() + 1}`.padStart(2, '0')
    const day = `${date.getDate()}`.padStart(2, '0')
    return `${year}-${month}-${day}`
  },

  adding(seconds: number, log: Record<string, number> | undefined, date: Date): Record<string, number> {
    const updated = { ...(log ?? {}) }
    const clamped = Math.max(0, Math.min(Math.trunc(seconds), this.sessionCapSeconds))
    if (clamped <= 0) return updated
    const key = this.dayKey(date)
    updated[key] = (updated[key] ?? 0) + clamped
    return updated
  },

  minutes(log: Record<string, number> | undefined, date: Date): number {
    return Math.floor((log?.[this.dayKey(date)] ?? 0) / 60)
  },
}

// MARK: - Spaced repetition

export const ReviewEngine = {
  newItem(exerciseID: string, now: Date): ReviewItem {
    return { id: exerciseID, exerciseID, due: new Date(now.getTime()), intervalDays: 0, ease: 2.3, repetitions: 0 }
  },

  recordSuccess(source: ReviewItem, now: Date): ReviewItem {
    const item = { ...source }
    item.repetitions += 1
    if (item.repetitions === 1) item.intervalDays = 1
    else if (item.repetitions === 2) item.intervalDays = 3
    else if (item.repetitions === 3) item.intervalDays = 7
    else item.intervalDays = Math.min(180, Math.round(Math.max(1, item.intervalDays) * item.ease))
    item.ease = Math.min(3, item.ease + 0.05)
    item.due = addDays(now, item.intervalDays)
    return item
  },

  recordFailure(source: ReviewItem, now: Date): ReviewItem {
    const item = { ...source }
    item.repetitions = 0
    item.intervalDays = 1
    item.ease = Math.max(1.6, item.ease - 0.2)
    item.due = addDays(now, 1)
    return item
  },
}
